crates/bevy_ecs/src/storage/sparse_set.rs

```diff
@@ -137,82 +139,84 @@ impl ComponentSparseSet {
     /// The `value` pointer must point to a valid address that matches the [`Layout`](std::alloc::Layout)
     /// inside the [`ComponentInfo`] given when constructing this sparse set.
     pub unsafe fn insert(&mut self, entity: Entity, value: OwningPtr<'_>, change_tick: u32) {
-        if let Some(&dense_index) = self.sparse.get(entity) {
-            self.dense.replace_unchecked(dense_index, value);
-            *self.ticks.get_unchecked_mut(dense_index) =
+        if let Some(&dense_index) = self.sparse.get(entity.id()) {
+            self.dense.replace_unchecked(dense_index as usize, value);
+            *self.ticks.get_unchecked_mut(dense_index as usize) =
                 UnsafeCell::new(ComponentTicks::new(change_tick));
         } else {
             let dense_index = self.dense.len();
             self.dense.push(value);
-            self.sparse.insert(entity, dense_index);
+            self.sparse.insert(entity.id(), dense_index as u32);
             debug_assert_eq!(self.ticks.len(), dense_index);
             debug_assert_eq!(self.entities.len(), dense_index);
             self.ticks
                 .push(UnsafeCell::new(ComponentTicks::new(change_tick)));
-            self.entities.push(entity);
+            self.entities.push(entity.id());
         }
     }
 
     #[inline]
     pub fn contains(&self, entity: Entity) -> bool {
-        self.sparse.contains(entity)
+        self.sparse.contains(entity.id())
```

Comment:
Here ignoring the generation is almost definitively incorrect.

Reply:
The SparseSetIndex of Entity already only checked the ID. The only change here is that we aren't saving an unused upper 4 bytes.

Reply:
This is definitely something we should communicate better, i.e. maybe an `AliveEntity` type :thinking: not for this PR though

Reply:
Perhaps a type alias that is split on debug assertions.